lib/dns/dns.cc

```diff
@@ -0,0 +1,1114 @@
+#include <endianness.hh>
+#include <errno.h>
+#include <thread.h>
+#include <tick_macros.h>
+#include <unwind.h>
+
+using Debug = ConditionalDebug<false, "DNS Resolver">;
+
+#include "../firewall/firewall.hh"
+
+#include "dns.hh"
+#include "protocol-headers.hh"
+
+/**
+ * This is an isolated stub DNS resolver for the CHERIoT network stack.
+ *
+ * This resolver supports A, AAAA, and CNAME queries on IPv4, for recursive DNS
+ * servers. We assume that the recursive resolver recurses into CNAME records.
+ *
+ * Support for IPv6 is left for future works.
+ *
+ * Since the resolver plugs directly with the firewall, it needs to know its
+ * own IP address, the IP address of the DNS server, its own MAC address, as
+ * well as the MAC address of the DNS server (or that of the gateway if the
+ * server is outside of the local network). The MAC address of the device is
+ * obtained from the firewall. The IP address of the device, of the DNS server,
+ * and the MAC address of the server/gateway are obtained from DHCP and ARP,
+ * whose corresponding packets are also forwarded to this resolver.
+ */
+
+namespace
+{
+	/**
+	 * Full ARP packet, assembled from Ethernet and ARP headers.
+	 */
+	struct FullARPPacket
+	{
+		EthernetHeader ethernet;
+		ARPHeader      arp;
+	} __packed;
+
+	/**
+	 * Full DNS packet, assembled from Ethernet, IPv4, UDP, and DNS
+	 * headers. Questions and answers are not included here (they have a
+	 * variable length) and follow the DNS header.
+	 */
+	struct FullDNSPacket
+	{
+		EthernetHeader ethernet;
+		IPv4Header     ipv4;
+		UDPHeader      udp;
+		DNSHeader      dns;
+	} __packed;
+
+	/**
+	 * Internal state of the DNS resolver.
+	 */
+	enum ResolverState : uint32_t
+	{
+		Uninitialized = 0,
+
+		// Initialization intermediate states
+		DeviceMACSet    = 1 << 0,
+		DNSServerMACSet = 1 << 1,
+		DeviceIPSet     = 1 << 2,
+		ServerIPSet     = 1 << 3,
+
+		// Ready to process requests
+		Ready = DeviceMACSet | DNSServerMACSet | DeviceIPSet | ServerIPSet,
+
+		// Waiting for an answer from the DNS server
+		WaitingForDNSReply = (1 << 4) | Ready,
+
+		// We got a successful answer, processing it
+		ProcessingDNSReply = (1 << 5) | Ready,
+
+		// The server returned an error, processing it
+		LookupFailed   = (1 << 6) | Ready,
+		LookupTimedOut = (1 << 7) | Ready
+	};
+	std::atomic<uint32_t> state = ResolverState::Uninitialized;
+
+	/**
+	 * When the resolver is waiting for an answer, the ID of the pending
+	 * query is stored in `queryID` by the caller thread. The firewall
+	 * thread will compare incoming DNS packets' identifiers with
+	 * `queryID`, and store matching answers in `queryResult` before
+	 * updating the `state`.
+	 */
+	uint16_t       queryID     = {0};
+	NetworkAddress queryResult = {0};
+
+	/**
+	 * MAC address of the device. We obtain this from the firewall.
+	 *
+	 * This is reset-critical: if corrupted, this will prevent the
+	 * compartment to recover from a crash.
+	 */
+	MACAddress deviceMAC = {0};
```

Comment:
I think that it would make sense. It would require some non-trivial changes in the firewall though - since this PR is already quite chunky I would prefer to mark this future work and open a bug report.

Reply:
(And I am happy to later PR & close that bug report)

Reply:
I would also add a comment to that variable to note that we could remove the "reset critical" label like that.